common/lib/common-utils/src/fluidError.ts

```diff
@@ -0,0 +1,66 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    IFluidError,
+    ITelemetryProperties,
+    ISensitiveDebugData,
+    isIFluidError,
+} from "@fluidframework/common-definitions";
+
+export abstract class FluidError extends Error implements IFluidError {
+    public abstract errorType: string;
+
+    private props: ITelemetryProperties = {};
+    private debugData: ISensitiveDebugData = {};
+    constructor(
+        message: string,
+        props: ITelemetryProperties = {},
+        debugData: ISensitiveDebugData = {},
+    ) {
+        super(message);
+        this.addDetails(props, debugData);
+    }
+
+    public getFluidTelemetryProps() { return { ...this.props, message: this.message, errorType: this.errorType }; }
+    public getSensitiveDebugData() {
+        return { ...this.debugData, stack: this.stack ?? "" };
+    } //* Or implement deep copy? Not even possible...?
+
+    public addDetails(props: Partial<ITelemetryProperties>, debugData: Partial<ISensitiveDebugData>) {
+        this.props = { ...this.props, ...props };
+        this.debugData = { ...this.debugData, ...debugData };
+    }
+}
+
+export class GenericFluidError extends FluidError {
+    public errorType: string = "generic";
+}
+
+export class ExternalFluidError extends FluidError {
```

Comment:
Maybe it will become clearer later in PR, but how do you handle cases where error is "inner", but we want to include "outer" properties to it? I.e. driver catches some error from fetch, but wants to provide its perspective on the error (like context of operation it was trying to do)?

To me it's one and the same, the only difference is that properties have different PII classification.
Maybe other way to say it - the primary pivot (errorType) is not internal vs. external, but type of error (the value from OdspErrorType | DriverError | ContainerErrorType | ...), the rest is just property annotation.